File: hardware/platformio/platformio.py
# Define the NOR gate
def NOR(a, b):
    return int(not (a or b))

# Logic circuit function
def logic_circuit(P, Q, R):
    A = NOR(P, Q)       # A = ¬(P + Q)
    B = NOR(Q, R)       # B = ¬(Q + R)
    C = NOR(P, R)       # C = ¬(P + R)
    D = NOR(Q, R)       # D = ¬(Q + R)
    
     # Stage 2
    X = NOR(A, B)
    Y = NOR(C, D)
    
    # Stage 3
    F = NOR(X,Y)
    return F

# Test all combinations of inputs P, Q, R
print("P Q R -> F")
for P in [0, 1]:
    for Q in [0, 1]:
        for R in [0, 1]:
            F = logic_circuit(P, Q, R)
            print(f"{P} {Q} {R} -> {F}")

File: hardware/platformio/platformio.c
#include <stdio.h>

// Define NOR gate function
int NOR(int a, int b) {
    return !(a || b);  // Logical NOR
}

// Logic circuit function
int logic_circuit(int P, int Q, int R) {
    int A = NOR(P, Q);   // A = ¬(P + Q)
    int B = NOR(Q, R);   // B = ¬(Q + R)
    int C = NOR(P, R);   // C = ¬(P + R)
    int D = NOR(Q, R);   // D = ¬(Q + R)

    // Stage 2
    int X = NOR(A, B);
    int Y = NOR(C, D);

    // Stage 3
    int F = NOR(X, Y);

    return F;
}

int main() {
    int P, Q, R;
    printf("P Q R -> F\n");

    for (P = 0; P <= 1; P++) {
        for (Q = 0; Q <= 1; Q++) {
            for (R = 0; R <= 1; R++) {
                int F = logic_circuit(P, Q, R);
                printf("%d %d %d -> %d\n", P, Q, R, F);
            }
        }
    }

    return 0;
}

